packages/itwinui-react/src/core/utils/components/ShadowRoot.tsx

```diff
@@ -68,19 +38,90 @@ export const ShadowRoot = ({
     return null;
   }
 
-  return (
-    <>
-      {shadowRoot ? (
-        ReactDOM.createPortal(children, shadowRoot)
-      ) : (
-        <template ref={attachShadowRef} />
-      )}
-    </>
+  return <ClientShadowRoot css={css}>{children}</ClientShadowRoot>;
+};
+
+// ----------------------------------------------------------------------------
+
+const ClientShadowRoot = ({ children, css }: ShadowRootProps) => {
+  const templateRef = React.useRef<HTMLTemplateElement>(null);
+  const shadowRoot = useShadowRoot(templateRef, { css });
+
+  // fallback to <style> tag if adoptedStyleSheets is not supported
+  const fallbackCss =
+    !supportsAdoptedStylesheets && css ? <style>{css}</style> : null;
+
+  return shadowRoot ? (
+    ReactDOM.createPortal(
+      <>
+        {fallbackCss}
+        {children}
+      </>,
+      shadowRoot,
+    )
+  ) : (
+    <template ref={templateRef} />
   );
 };
 
 // ----------------------------------------------------------------------------
 
+/**
+ * Given a ref, this hook will return a shadowRoot attached to its parent element.
+ *
+ * The css will be added to the shadowRoot using `adoptedStyleSheets` (if supported).
+ */
+function useShadowRoot(
+  templateRef: React.RefObject<HTMLElement>,
+  { css = '' },
+) {
+  const [shadowRoot, setShadowRoot] = React.useState<ShadowRoot | null>(null);
+  const styleSheet = React.useRef<CSSStyleSheet>();
+  const latestCss = useLatestRef(css);
+
+  useLayoutEffect(() => {
+    const parent = templateRef.current?.parentElement;
+    if (!parent) {
+      return;
+    }
+
+    if (parent.shadowRoot) {
+      parent.shadowRoot.replaceChildren(); // Remove previous shadowroot content
+    }
+
+    const shadow = parent.shadowRoot || parent.attachShadow({ mode: 'open' });
+
+    if (supportsAdoptedStylesheets) {
+      // create an empty stylesheet and add it to the shadowRoot
+      const _global = shadow.ownerDocument.defaultView || globalThis;
+      styleSheet.current = new _global.CSSStyleSheet();
+      shadow.adoptedStyleSheets = [styleSheet.current];
+
+      // add the CSS immediately to avoid FOUC (one-time)
+      if (latestCss.current) {
+        styleSheet.current.replaceSync(latestCss.current);
+      }
+    }
+
+    queueMicrotask(() => {
+      ReactDOM.flushSync(() => setShadowRoot(shadow));
+    });
```

Comment:
A small comment explaining why this is wrapped in a `flushSync`/`queueMicrotask` would be helpful.

Reply:
this good?

```suggestion
    queueMicrotask(() => {
      // Flush the state immediately to prevent flash of unslotted light DOM
      ReactDOM.flushSync(() => setShadowRoot(shadow));
    });
```

Reply:
Yes, that sounds good for the `flushSync()`. But I'm also just wondering, what is the reason for wrapping it in a `queueMicrotask`?

Reply:
you can't `flushSync` inside an Effect. i don't think that needs a comment, because React's error is clear enough when you try it

Reply:
Updated the comment a little bit, because `flushSync` doesn't actually prevent the flash of unslotted content. [`c783d4e` (#1954)](https://github.com/iTwin/iTwinUI/pull/1954/commits/c783d4e9bcbba8f9b7e1a6a25062a072b794ba20)

To understand this more deeply, you can comment out the `flushSync` and call `setShadowRoot`, then visit `Dialog` stories.

Reply:
To confirm, which `Dialog` stories exactly are you suggesting me to visit? And what should I keep an eye out for?

Reply:
Actually I can't reproduce it anymore. But this `flushSync` was added in #1836, where I noticed that the measurements used by the draggable Dialog here were wrong:

https://github.com/iTwin/iTwinUI/blob/e3bdda14a320540036ee04a706c1b4d19783df03/packages/itwinui-react/src/core/Dialog/DialogMain.tsx#L139-L143